neuralhydrology/evaluation/metrics.py

```diff
@@ -625,6 +625,45 @@ def mean_peak_timing(obs: DataArray,
     return np.mean(timing_errors) if len(timing_errors) > 0 else np.nan
 
 
+def peak_flow_bias(obs: DataArray,
+                   sim: DataArray) -> float:
+    """Calculate the percent bias of peak flows.
+
+    Uses scipy.find_peaks to find peaks in the observed time series. The observed peaks indices are used to subset
+    observed and simulated flows. Finally, the peak flow bias is calculated as the percent bias of observed peak flows
+    and corresponding simulated flows.
+
+    Parameters
+    ----------
+    obs : DataArray
+        Observed time series.
+    sim : DataArray
+        Simulated time series.
+
+    Returns
+    -------
+    float
+        Peak flow bias in percent.
+    """
+    # verify inputs
+    _validate_inputs(obs, sim)
+
+    # get time series with only valid observations
+    obs, sim = _mask_valid(obs, sim)
+
+    # heuristic to get indices of peaks and their corresponding height.
+    peaks, _ = signal.find_peaks(obs.values, distance=100, prominence=np.std(obs.values))
+
+    # subset data to only peak values
+    obs = obs[peaks]
+    sim = sim[peaks]
+
+    # calculate the peak error
+    peak_error = np.sum(sim - obs) / np.sum(obs)
```

Comment:
I think this is open for discussion but I thought more about something like the "mean absolute peak flow bias (in percent)", rather than the overall peak bias.

IIuc, in your equation, over- and underestimates of different peaks could cancel each other out (because of the missing `np.abs()`, leading potentially to a perfect peak-bias score even though there are errors. Also, by summing over all errors first (and all obs), we would neglect that absolute errors on different magnitudes are in different value ranges.

E.g. imagine we have two peaks, one with obs=2 and one with obs=200. For the first peak, imagine sim=1 and for the second sim=180. According to your equation, the peak_flow_bias would be ~10% (21/202 * 100). If you would compute the percentage bias on each peak first, and then average across peaks, you would get (1/2 + 20/200)/2 * 100 = 30%.

Since this PR is proposing a new metric, there is no right or wrong, but intuitively (also considering the other metrics here, like the peak timing metric), I would find the "mean peak bias in percent" clearer. @gauchm @danklotz opinions?

Reply:
The equation is inspired by the $\text{BiasFHV}$ equation, which also does not take into account absolute differences in simulated and observed flows. I see that there is a shortcoming, resulting from the coincident of postive and negative differences, which eliminates each other. However, if we take the absolute differences for calculation, the resulting metric does not say anaything about over- or underestimates of the peaks.

Regarding the question about considering the overall or mean peak bias, I'm not sure which approach yields a more valuable accuracy metric. Sure, taking the mean of all differences is much clearer and each peak flow difference is weighted equally. However, differences in very high peaks have a greater impact on the result than differences in lower peaks. My feeling is, that a difference of e.g. 10 m³/s in a very high peak is less relevant than the same difference in very low peaks. Though, normalizing the difference by the sum of all observed flows, such as in the $\text{BiasFHV}$, adresses this issue.

So, as you already mentioned, a discussion about this new metric would be very helpful to clear up its understanding.

Reply:
To be honest, I don't have much of an opinion on this issue - sorry. For me both the abs and the non-abs approach are totally fine.
I think I haven't fully understood the argument about overall vs. mean peak bias. Isn't the impact of the same difference for a high peak _lower_ (and not higher) than for a low peak (since Freddy wants to divide by the corresponding obs)?

Reply:
>Isn't the impact of the same difference for a high peak lower (and not higher) than for a low peak (since Freddy wants to divide by the corresponding obs)?

Yes, absolutely. Actually, that is what I meant. The argument regarding higher and lower impact relates to the mean of all differences. Taking the mean will score all differences equally. So differences in high peaks, which are likely to be also high, will impact the mean more than differences in lower peaks, which are likely to be lower. But instead, as you also mention, differences in lower peaks should have an higher impact than the same differences in higher peaks.

Reply:
As Martin I also have no strong feelings about this. Here are my thoughts: 

If we want to call the metric "xyz bias" then I would for using the version without the absolute. My reasoning is that it then does more closely correspond to what math people would expect from a bias (in the sence that we check if the expectations do correspond). 

From a diagnostic standpoint, however, I like the version with the absolute better. As you say, we loose information about wheter we over- or underestimate (which imo is also not well captured by the current bias), but we would be much sharper with respect to the error magnitudes. 

Since over- and underestimation is important we might still want to recuperate that info at some point (maybe yet another metric?).

Reply:
So to summarize, 3 out of 4 think the metric as is is fine? Then I vote for merging the metric into master. We can always add other metrics and/or modify the metric later. 

Thanks again @SebaDro for your effort here. I will wait for feedback until tomorrow, otherwise I will merge it. And sorry again for my delayed replies but I am currently on vacation and only have a quick look at the internet in the evening.

Reply:
I think it is a 2 ( @kratzert + @danklotz ) vs 2 ( @gauchm + @SebaDro ) currently. 

Just to make my position clear: I prefer the version of @kratzert (The one with the `np.abs`). The reason for that is that it seems to me the better diagnostic tool. The other version can compensate errors (a) between peaks and (b) between basins. We loose some info about over and under estimation, but imo gain more.

I just would like to call it differently; say *relative mean absolute peak error*.

Reply:
Actually, I'd be fine with both versions. Although I introduced the non-abs version, I also think that taking the absolute error gives more information about the model performance in peak regions. So, it's totally ok, if we vote for that metric. 

Though, there is still one decision pending regarding taking the mean absolute error or the overall percent bias? As far as I can see, there are three different implementation and metric name options:
1. Mean absolute peak error (MAE_peak):
$\text{MAE}\_\text{peak} =\frac{1}{P} \sum_{p=1}|Q_{s,p} - Q_{o,p}|$
```python
mae_peak = np.sum(np.abs(sim - obs)) / peaks.size()
```

2. Absolute percentage peak error (APE_peak):
$\text{APE}\_\text{peak} = \frac{\sum_{p=1}|Q_{s,p} - Q_{o,p}|}{\sum_{p=1}Q_{o,p}} \times 100$
```
ape_peak = np.sum(np.abs(sim - obs)) / np.sum(obs) * 100
```

3. Mean absolute percentage peak error (MAPE_peak):
$\text{MAPE}\_\text{peak} =  \frac{1}{P}\sum_{p=1}\frac{|Q_{s,p} - Q_{o,p}|}{Q_{o,p}}\times 100$
```
mape_peak = np.sum(np.abs(sim - obs) / np.sum(obs)) / peaks.size() * 100
```

I'd vote for option three, since it is a well known metric for percentage prediction errors in statistics. There also exists a [scikit-learn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_absolute_percentage_error.html) for it. So, I'd assume that there should be a common understanding of this metric. 

Just give a vote for one of the options and I will update the pull request in accordance to the voting result.

Reply:
Sorry if I am annoying, but from what I gather MAPE seems to be defined slightly different (see [a](https://en.m.wikipedia.org/wiki/Mean_absolute_percentage_error),[b](https://github.com/scikit-learn/scikit-learn/blob/36958fb24/sklearn/metrics/_regression.py#L296)). Apart from that I am fine with 2, 3, or other MAPE versions ☺️

Reply:
Oh yes, thanks. In fact, the above shown MAPE implementation does not fit to the formula. Actually (assuming, observations are always positive), it should be:
```
mape_peak = np.sum(np.abs(sim - obs) / obs) / peaks.size() * 100
```
 But to fit the original formula, we also could write :
```
mape_peak = np.sum(np.abs((sim - obs) / obs)) / peaks.size() * 100
```
The formula then would be $\text{MAPE}\_\text{peak} =  \frac{1}{P}\sum_{p=1}|\frac{Q_{s,p} - Q_{o,p}}{Q_{o,p}}|\times 100$, with $\text{P}$ as the number of peaks considered.

Reply:
+1 looks good to me.

Reply:
+1 from me as well. I'd say let's go ahead with this version?

Reply:
+1

Reply:
Great 👍. I'll come with an updating commit as soon as possible.